base/indices.jl

```diff
@@ -357,6 +357,14 @@ LinearIndices(inds::NTuple{N,Union{<:Integer,AbstractUnitRange{<:Integer}}}) whe
     LinearIndices(map(i->first(i):last(i), inds))
 LinearIndices(A::Union{AbstractArray,SimpleVector}) = LinearIndices(axes(A))
 
+promote_type(::Type{NTuple{N,R1}}, ::Type{NTuple{N,R2}}) where {N,R1<:AbstractUnitRange,R2<:AbstractUnitRange} =
+    NTuple{N,promote_type(R1,R2)}
```

Comment:
What makes tuples of unit ranges special compared to other tuples to justify this? (Other than that this comes handy in this PR?)

Reply:
Fair enough. But it's not wrong either. However, I could hide it behind a private method if you really think that's better.

Reply:
This method is running into some trouble too with [ambiguity (really, typevar) tests](https://github.com/JuliaLang/julia/blob/b0f531e5f0b626f1ee91bb7cac2a0bb660f435f3/test/ambiguous.jl#L272-L303). But I think it's a bug, see #28086.

Reply:
I'd lean towards making this a private method, too.

Reply:
Yeah. For fun I played with making it really general,
```diff
index 78a1954..f260cbf 100644
--- a/base/promotion.jl
+++ b/base/promotion.jl
@@ -236,6 +236,11 @@ promote_rule(::Type{Any}, ::Type{<:Any}) = Any
 promote_rule(::Type{<:Any}, ::Type{Any}) = Any
 promote_rule(::Type{Any}, ::Type{Any}) = Any
 
+function promote_rule(::Type{Tuple{R1,Vararg{R1,N}}}, ::Type{Tuple{R2,Vararg{R2,N}}}) where {R1,R2,N}
+    R = promote_type(R1, R2)
+    Tuple{R,Vararg{R,N}}
+end
+
```
Somewhat surprisingly, only one test fails: https://github.com/JuliaLang/julia/blob/master/test/abstractarray.jl#L653
But I'll make it a private method, because that really would be a scary change.